### PR TITLE
docs/network: First step on describing standard network class interface.

### PR DESCRIPTION
**Pull request by pfalcon**
Using BaseNIC "base class".

### CODE REVIEW

**Comment by pfalcon** [comment]
First cut of work for https://github.com/micropython/micropython/issues/2998

**Comment by pfalcon** [comment]
Ok, no concerns raised, which is good. So, I while definitely not complete, I propose to merge this to have further progress. @dpgeorge

File: docs/library/network.rst [hunk 1]
```diff
@@ -29,6 +29,95 @@ For example::
...
+       sets of predefined/recommended paramaters, among them:
+
+       * WiFi: `bssid` keyword to connect by BSSID (MAC address) instead
+         of access point name
+
+    .. method:: disconnect()
+
+       Disconnect from network.
+
+    .. method:: isconnected()
```
Comment (tomlogic) [inline docs/library/network.rst]: I realize we have legacy with the `isconnected()` name, but wouldn't `is_connected()` match other naming conventions, including PEP8?

File: docs/library/network.rst [hunk 2]
```diff
@@ -29,6 +29,95 @@ For example::
...
+       for services of a particular set, etc.), and to affect scan
+       duration and other parameters. Where possible, parameter names
+       should match those in connect().
+
+    .. method:: status()
+
+       Return detailed status of the interface, values are dependent
+       on the network medium/technology.
+
+    .. method:: ifconfig([(ip, subnet, gateway, dns)])
```
Comment (tomlogic) [inline docs/library/network.rst]: I've been working on a platform with cellular connectivity, running MicroPython in a sandbox.  My recommendation is for ifconfig() to return a dict including any number of predefined keys that we document here.  For example, a cellular connection is point-to-point and doesn't have a subnet.  Should `dns` be a single IP address, or can it return a list of addresses?  Would someone want to know their MAC address?  With cellular, we're returning the device's phone number and information on the modem/SIM.  Accept a dict of parameters to set, so you can leave remaining parameters unchanged.

File: docs/library/network.rst (hunk 1 above)
Comment (pfalcon) [inline docs/library/network.rst]: Nope. There're few different naming conventions used with Python, one used within core avoids using extra underscores, e.g. `isinstance()`. This module/method is arguably close to the core and  *follows* this convention (not random), so there're no point in changing it. (Of course, user libraries are actually discouraged to use this "core" naming convention.)

File: docs/library/network.rst (hunk 2 above)
Comment (pfalcon) [inline docs/library/network.rst]: ifconfig() is legacy, fixed-function (ahem) function. Core interface to to the underlying network interface (ahem again) is the extensible (within bounds defined, no dicts returned and stuff) .config() function. More general control of IP networking (DHCP, DNS) is yet to be found.

Comment (tomlogic) [inline docs/library/network.rst]: Thanks for the feedback Paul.  I'll work on implementing the standard interface with our codebase, and a proposal for ways to extend it if I find myself stuck.

How about documenting that `ifconfig()` could return `None` as one of the IP addresses if it doesn't apply or isn't available?  Or would you expect `'0.0.0.0'` in that case?  If a user wants to set the DNS server without changing existing settings, I suppose they would do something like:

    c = nic.ifconfig()
    nic.ifconfig((c[0], c[1], c[2], '8.8.4.4'))  # change DNS without modifying other settings

I would propose documenting (and supporting) using `None` as placeholders in the tuple for elements that you don't want to change.

    nic.ifconfig((None, None, None, '8.8.4.4'))  # change DNS without modifying other settings

Comment (pfalcon) [inline docs/library/network.rst]: So, work is being done here is two-fold:

* Find a "union" of the functionality already implemented in existing network modules.
* Define a "superset" of that functionality, as much as possible in a compatible way.

So, "None as placeholders in the tuple for elements that you don't want to change" - is it implemented for any of the mainline platforms? If not, what's the point of documenting it if doesn't work?

Beyond that, what if a user wants to "undefine" gateway or DNS? "None" is the natural value for that, yet you propose to assign it a special meaning.

This shows that designing APIs a tricky matter. While I empathize with the idea to make it "easier", I'm afraid using your 2-liner is the best option we have - now and for the foreseeable future. Again, the idea is first to try to capture what we already have implemented, e.g. things which were implemented for esp8266, but not documented. Then, things which users want to be implemented for esp8266 (some of these things contradict one another!).

File: docs/library/network.rst [hunk 3]
```diff
@@ -29,6 +29,95 @@ For example::
...
+       of a service to connect to. It may be accompanied by a key
+       (password) required to access service. There can be further
+       arbitrary keyword-only parameters, depending on networking medium
+       type and/or particular device. Parameters can be used to: a)
+       specify alternative service identifer types; b) provide additional
+       connection parameters. For various medium types, there are different
+       sets of predefined/recommended paramaters, among them:
+
+       * WiFi: `bssid` keyword to connect by BSSID (MAC address) instead
+         of access point name
```
Comment (dpgeorge) [inline docs/library/network.rst]: It's not 100% clear here if the ssid (service id) is ignored if bssid is given.  I guess that may even be an implementation detail for a given NIC.

File: docs/library/network.rst [hunk 4]
```diff
@@ -29,6 +29,95 @@ For example::
...
+
+    .. method:: disconnect()
+
+       Disconnect from network.
+
+    .. method:: isconnected()
+
+       Returns True if connected to network.
```
Comment (dpgeorge) [inline docs/library/network.rst]: "Returns `True` if connected to a network, otherwise returns `False`.'

File: docs/library/network.rst [hunk 5]
```diff
@@ -29,6 +29,95 @@ For example::
...
+
+    .. method:: scan(\*, ...)
+
+       Scan for the available network services/connections. Returns
+       list of tuples with discovered service parameters. For various
```
Comment (dpgeorge) [inline docs/library/network.rst]: "Returns a list of ..."

File: docs/library/network.rst [hunk 6]
```diff
@@ -29,6 +29,95 @@ For example::
...
+       network media, there are different variants of predefined/
+       recommended tuple formats, among them:
+
+       * WiFi: (ssid, bssid, channel, RSSI, authmode, hidden). There
+       may be further fields, specific to a particular device.
+
+       The may accept additional keyword arguments to filter scan results
```
Comment (dpgeorge) [inline docs/library/network.rst]: "The function may ..."

File: docs/library/network.rst [hunk 7]
```diff
@@ -29,6 +29,95 @@ For example::
...
+        nic.ifconfig(('192.168.0.4', '255.255.255.0', '192.168.0.1', '8.8.8.8'))
+
+    .. method:: config('param')
+                config(param=value, ...)
+
+       Get or set general network interface parameters. These methods allow to work
+       with additional parameters beyond standard IP configuration (as dealt with by
+       ``ifconfig()``). These include network-specific and hardware-specific
+       parameters and status values. For setting parameters, keyword argument syntax
+       should be used, multiple parameters can be set at once. For querying, a
```
Comment (dpgeorge) [inline docs/library/network.rst]: "... the keyword argument syntax should be used, and multiple ..."

**Review by dpgeorge — COMMENTED** [review]
It may be worth making explicit in the intro paragraph that the network module is for IP-based networks.  Well, maybe that's something that needs discussion, but will impact, eg, whether bluetooth/cellular/lora/etc interfaces are added to this module.

File: docs/library/network.rst (hunk 3 above)
Comment (pfalcon) [inline docs/library/network.rst]: Yep, if the behavior is not defined, it's undefined ;-).

File: docs/library/network.rst (hunk 7 above)
Comment (tomlogic) [inline docs/library/network.rst]: Assumptions I'm making while exploring implementation of this BaseNIC spec.  If they're invalid assumptions, we might need to update the documentation.

- The documentation doesn't specify whether `config()` returns anything when called with keywords, so I would expect it to return `None`.  
- I would expect a `TypeError` thrown on an invalid parameter, before any parameters were actually applied.
- We should explicitly call out whether parameters are set in the order passed.  If a user needs a specific order, they should probably make separate calls to `config()`.
- I would expect an exception if I tried to set one of the parameters to an invalid value, and that some, all or none of the other parameters may have been set before the exception.

Comment (pfalcon) [inline docs/library/network.rst]: > The documentation doesn't specify whether config() returns anything when called with keywords, so I would expect it to return None.

That's true. I'm sure CPython documentation doesn't list explicit return value for a lot of setter functions, as None is implied.

> I would expect a TypeError thrown on an invalid parameter, before any parameters were actually applied.

As in: name of the parameter? Perhaps raising TypeError would make sense, to follow behavior of "normal" keyword functions. However, currently esp8266 raises ValueError. However, "before any parameters were actually applied" is non-goal. All parameters are processed once in one loop, applied immediately and in row. That should be the expected behavior.

> We should explicitly call out whether parameters are set in the order passed.

I believe anything like that was impossible at all before CPython 3.6. Consider it underspecified for .config() call.

> If a user needs a specific order, they should probably make separate calls to config().

Yes.

> I would expect an exception if I tried to set one of the parameters to an invalid value, and that some, all or none of the other parameters may have been set before the exception.

Yes, with addition that invalid parameter name leads to the same effect, per above.

File: docs/library/network.rst [hunk 8]
```diff
@@ -29,6 +29,95 @@ For example::
...
+
+.. class:: BaseNIC(id=None, ...)
+
+Create a network interface object. Parameters are network interface
+dependent. If there are few interfaces of the same type, the first
+parameter should be `id`.
+
+    .. method:: active([is_active])
+
+        Activate ("up") or deactivate ("down") network interface, if boolean
```
Comment (pfalcon) [inline docs/library/network.rst]: fixed

File: docs/library/network.rst [hunk 9]
```diff
@@ -29,6 +29,95 @@ For example::
...
+        argument is passed. Otherwise, query current state if no argument is
+        provided. Most other methods require an active interface.
+
+    .. method:: connect([service_id, key=None, \*, ...])
+
+       Connect interface to the network. This method is optional, and
```
Comment (pfalcon) [inline docs/library/network.rst]: fixed

File: docs/library/network.rst [hunk 10]
```diff
@@ -29,6 +29,95 @@ For example::
...
+       available only for interfaces which are not "always connected".
+       If no parameters are given, connect to the default (or only)
+       service. If a single parameter is given, it is primary identifier
```
Comment (pfalcon) [inline docs/library/network.rst]: fixed

File: docs/library/network.rst [hunk 11]
```diff
@@ -29,6 +29,95 @@ For example::
...
+       of a service to connect to. It may be accompanied by a key
+       (password) required to access service. There can be further
```
Comment (pfalcon) [inline docs/library/network.rst]: fixed

File: docs/library/network.rst [hunk 12]
```diff
@@ -29,6 +29,95 @@ For example::
...
+       arbitrary keyword-only parameters, depending on networking medium
```
Comment (pfalcon) [inline docs/library/network.rst]: fixed

File: docs/library/network.rst [hunk 13]
```diff
@@ -29,6 +29,95 @@ For example::
...
+       type and/or particular device. Parameters can be used to: a)
+       specify alternative service identifer types; b) provide additional
+       connection parameters. For various medium types, there are different
+       sets of predefined/recommended paramaters, among them:
```
Comment (pfalcon) [inline docs/library/network.rst]: fixed

File: docs/library/network.rst (hunk 4 above)
Comment (pfalcon) [inline docs/library/network.rst]: Fixed, but I wonder if that's too verbose, we need to check whether CPython docs say it like that in all cases.

File: docs/library/network.rst (hunk 5 above)
Comment (pfalcon) [inline docs/library/network.rst]: fixed

File: docs/library/network.rst (hunk 6 above)
Comment (pfalcon) [inline docs/library/network.rst]: fixed

File: docs/library/network.rst (hunk 7 above)
Comment (pfalcon) [inline docs/library/network.rst]: fixed

**Comment by pfalcon** [comment]
> It may be worth making explicit in the intro paragraph that the network module is for IP-based networks. Well, maybe that's something that needs discussion, but will impact, eg, whether bluetooth/cellular/lora/etc interfaces are added to this module.

Yes, let's discuss. De-facto, this module was used to configure IP networking devices, its usage example told "configure NIC using network module, then use socket module as usual". The point, that network module provided only configuration facilities, and "application" layer was in another, standard module. 

CC3200 port was the only port which went above and beyond that, adding half-system, half-application confusingly named "Server" class there, which actually "controls the behaviour and the configuration of the FTP and telnet services running on the WiPy." As a next step, I'm removing description of it from network.rst, then perhaps that class should be moved to "wipy" or some other module.

Summing up: "network" module is for configuring networking params, not for application-level things. If we allow non-IP networking classes into "network", the question is where application level for them will be. Stuffing everything into "network" is too much, and if there will be a separate module for a particular networking tech, it makes sense to have everything for that tech (both config and app layer) in one module. Especially that for adhoc (non-IP) networking, config and app layers are usually more tightly coupled.

On the other hand, "network" module defines few meta-patterns, which are highly recommended for reuse by other modules, e.g. .config() call which allows to either configure or query arbitrary params. Actually, I'd consider it very fruitful pattern, which should be reused in as many modules as possible, beyond just networking. For example, I remember there was a proposal (not mine) to replace hardware API's ("machine") .init() call with .config(). I actually have an idea to dig it out and to resubmit for consideration, based on the need to finalize Pin configuration interface.

**Comment by pfalcon** [comment]
The patch otherwise was updated: @dpgeorge's typo fixes applied, the class described is renamed from BaseNIC to AbstractNIC, to emphasize that it's not supposed to be used (or even seen) directly, extended use example and some method descriptions.

**Comment by pfalcon** [comment]
>  As a next step, I'm removing description of it from network.rst

92206a78aeddf81eca08fb25589231e5a62df69c

**Comment by pfalcon** [comment]
Merged in the meantime.

**Comment by dpgeorge** [comment]
> Stuffing everything into "network" is too much, and if there will be a separate module for a particular networking tech, it makes sense to have everything for that tech (both config and app layer) in one module. Especially that for adhoc (non-IP) networking, config and app layers are usually more tightly coupled.

Yes, some network technologies, like Bluetooth, do have a strong coupling between config and usage.  The original reason for "network" module was because CPython provides only "socket" and no way to configure the underlying interface.  This fact shows that there is good separation for IP networking between the config and the use.

Some technologies, eg cellular, BLE, may allow to have an IP network run on top of them and in this case it would make sense to put their configuration in "network".  But they'd also have other non-IP usage, with an API that would live somewhere else.

**Comment by dpgeorge** [comment]
This was merged in cd64b3082e501f5a54cb15ab3087be91c8f90338

**Pull request closed without merging**